sky/jobs/controller.py

```diff
@@ -379,18 +410,69 @@ def run(self):
+            else:
+                logger.info(f'Adding task {task_id} to failed tasks.')
+                self._task_status[task] = TaskStatus.FAILED
+                self._cancel_all_tasks(task_id)
+
+    def _cancel_all_tasks(self, task_id: int):
+        callback_func = managed_job_utils.event_callback_func(
+            job_id=self._job_id, task_id=task_id, task=self._dag.tasks[task_id])
+        for task in self._dag.tasks:
+            if task not in self._task_status:
+                self._task_status.setdefault(task, TaskStatus.CANCELLED)
+
+        # Call set_cancelling before set_cancelled to make sure the table
+        # entries are correctly set.
+        managed_job_state.set_cancelling(self._job_id, callback_func)
+        managed_job_state.set_cancelled(self._job_id, callback_func)
+
+    def run(self):
+        """Run controller logic and handle exceptions."""
+        all_tasks_completed = lambda: self._num_tasks == len(self._task_status)
+        # TODO(andy):Serve has a logic to prevent from too many services running
+        # at the same time. We should have a similar logic here, but instead we
+        # should calculate the sum of the subtasks (an upper bound), instead of
+        # the number of jobs (dags).
+        # Further, we could try to calculate the maximum concurrency in the dag
+        # (e.g. for a chain dag it is 1 instead of n), which could allow us to
+        # run more dags in parallel.
+        max_workers = self._num_tasks
+        managed_job_utils.make_launch_log_dir_for_redirection(self._job_id)
+        with futures.ThreadPoolExecutor(max_workers) as executor:
+            future_to_task = {}
+            while not all_tasks_completed():
+                while not self._task_queue.empty():
+                    task_id = self._task_queue.get()
+                    log_file_name = managed_job_utils.get_launch_log_file_name(
+                        self._job_id, task_id)
+
+                    logger.info(
+                        f'Task {task_id} is submitted to run. To see logs: '
+                        f'{ux_utils.BOLD}sky jobs logs {self._job_id} '
+                        f'--task-id {task_id}{ux_utils.RESET_BOLD}')
+
+                    logger.info(f'Redirecting output to {log_file_name}.')
+
+                    with ux_utils.RedirectOutputForThread() as redirector:
+                        future = executor.submit(
+                            redirector.run(self._run_one_task, log_file_name),
+                            task_id, self._dag.tasks[task_id])
+                    future_to_task[future] = task_id
+
+                done, _ = futures.wait(future_to_task.keys(),
+                                       return_when='FIRST_COMPLETED')
+
+                for future in done:
+                    logger.info(f'Task {future_to_task[future]} completed.')
+                    task_id = future_to_task.pop(future)
+                    self._handle_future_completion(future, task_id)
```

Comment:
```suggestion
                for future in done:
                    logger.info(f'Task {future_to_task[future]} completed.')
                    task_id = future_to_task.pop(future)
                    self._handle_future_completion(future, task_id)
                time.sleep(1)
```
should we add a sleep here to avoid it always running?

Reply:
`_ready_tasks.get` will block until there is new task ready. Refer to https://docs.python.org/3/library/queue.html#queue.Queue.get